File: 面试高频题/0040 组合总和 II.py
from typing import List

class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        def dfs(begin, path, remain):
            if remain == 0:
                res.append(path[:])
                return
            for i in range(begin, size):
                if candidates[i] > remain:
                    break
                if i > begin and candidates[i-1] == candidates[i]:
                    continue
                path.append(candidates[i])
                dfs(i+1, path, remain-candidates[i])
                path.pop()
        size = len(candidates)
        if size == 0:
            return []
        candidates.sort()
        res = []
        dfs(0, [], target)
        return res

File: 面试高频题/0040 组合总和 II.cpp
class Solution {
public:
    vector<vector<int>> res;
    vector<int> temp;

    void dfs(vector<int>& candidates, int target, int idx) {
        if (target==0) {
            res.push_back(temp);
            return ;
        }

        for (int i=idx; i<candidates.size(); ++i) {
            if (target - candidates[idx] < 0)
                break;
            if (i>idx && candidates[i] == candidates[i-1])
                continue;
            temp.push_back(candidates[i]);
            dfs(candidates, target-candidates[i], i+1);
            temp.pop_back();
        }
    }

    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        sort(candidates.begin(), candidates.end());
        dfs(candidates, target, 0);
        return res;
    }
};